### PR TITLE
Split nextpnr.h to allow for linear inclusion.

### PR DESCRIPTION
**Pull request by litghost**
"nextpnr.h" is no longer the god header.  Important improvements:

 - Functions in log.h can be used without including
   BaseCtx/Arch/Context. This means that log_X functions can be called
   without included "nextpnr.h"

 - NPNR_ASSERT can be used without including "nextpnr.h" by including
   "nextpnr_assertions.h".  This allows NPNR_ASSERT to be used safely in
   any header file.

 - Types defined in "archdefs.h" are now available without including
   BaseCtx/Arch/Context.  This means that utility classes that will be
   used inside of BaseCtx/Arch/Context can be defined safely in a
   self-contained header.

### CODE REVIEW

**Comment by gatecat** [comment]
Thanks, this looks like a great improvement. I will give it a proper review over the weekend.

**Comment by litghost** [comment]
> Thanks, this looks like a great improvement. I will give it a proper review over the weekend.

I suspect I'll need to fixup many of the arches.

**Comment by litghost** [comment]
> > Thanks, this looks like a great improvement. I will give it a proper review over the weekend.
> 
> I suspect I'll need to fixup many of the arches.

Build is passing now.

File: common/deterministic_rng.h [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include <algorithm>
```
Comment (gatecat) [inline common/deterministic_rng.h:21]: nitpick but I think these should be after the include guard

**Comment by gatecat** [comment]
I think with this PR places using `namespace nextpnr` like bits.h can now go back to `NEXTPNR_NAMESPACE_BEGIN` for consistency

**Comment by litghost** [comment]
> I think with this PR places using `namespace nextpnr` like bits.h can now go back to `NEXTPNR_NAMESPACE_BEGIN` for consistency

I agree, but I'd like to fix that in a follow up PR, rather than in this one.

File: common/deterministic_rng.h (hunk 1 above)
Comment (litghost) [reply common/deterministic_rng.h:21]: Will fix.

**Comment by gatecat** [comment]
> I agree, but I'd like to fix that in a follow up PR, rather than in this one.

No problem, just checking it wasn't forgotten about.

File: common/deterministic_rng.h (hunk 1 above)
Comment (litghost) [reply common/deterministic_rng.h:24]: Fixed

**Pull request merged**